taskvine/src/worker/vine_worker.c

```diff
@@ -1492,27 +1543,47 @@ static void check_libraries_ready()
 	library_link_info.events = LINK_READ;
 	library_link_info.revents = 0;
 
-	/* Loop through all processes to find libraries and check if they are ready. */
+	/* Loop through all processes to find libraries and check if they are alive. */
 	ITABLE_ITERATE(procs_running, library_task_id, library_process)
 	{
-		if ((library_process->type == VINE_PROCESS_TYPE_LIBRARY) && !library_process->library_ready) {
-			library_link_info.link = library_process->library_read_link;
+		if (library_process->type != VINE_PROCESS_TYPE_LIBRARY)
+			continue;
 
-			/* Check if library has sent its startup message. */
-			if (link_poll(&library_link_info, 1, 0) && (library_link_info.revents & LINK_READ)) {
-				if (check_library_startup(library_process)) {
-					debug(D_VINE,
-							"Library %s reports ready to execute functions.",
-							library_process->task->provides_library);
-					library_process->library_ready = 1;
-				} else {
-					/* Kill library if the name reported back doesn't match its name or
-					 * if there's any problem. */
-					debug(D_VINE, "Library verification failed. Killing it.");
-					vine_process_kill(library_process);
-				}
+		/* Check if library has sent its startup message. */
```

Comment:
Hmm, I am not sure that the heartbeat idea is actually making things better here.

The previous idea was that the library sends a startup message once it has completed its own internal initialization.  (And that could take longer than 30s.)

This is adding more pipes and things to check, but not actually improving upon that detection.

I think you have enough good things in this PR without adding the heartbeat mechanism.